source/extensions/filters/http/rbac/rbac_filter.h

```diff
@@ -39,18 +43,54 @@ class RoleBasedAccessControlFilterConfig {
       const std::string& stats_prefix, Stats::Scope& scope);
 
   RoleBasedAccessControlFilterStats& stats() { return stats_; }
+  const envoy::config::filter::http::rbac::v2::RBAC& config() { return config_; }
 
   const Filters::Common::RBAC::RoleBasedAccessControlEngine&
-  engine(const Router::RouteConstSharedPtr route) const;
+  engine(const Router::RouteConstSharedPtr route, EnforcementMode mode) const;
+
+  bool isEnabled(const Router::RouteConstSharedPtr route, EnforcementMode mode) const;
 
 private:
+  const Filters::Common::RBAC::RoleBasedAccessControlEngine& engine(EnforcementMode mode) const {
+    return mode == EnforcementMode::ENFORCED ? engine_ : shadow_engine_;
+  }
+
+  bool isEnabled(EnforcementMode mode) const {
+    return mode == EnforcementMode::ENFORCED ? config_.has_rules() : config_.has_shadow_rules();
+  }
```

Comment:
use two boolean member variables instead of holding config _?  we could do that that but personally I don't feel too much differences (run-time also should be quick ?)

Reply:
+1, you can make `engine_` and `shadow_engien_` an `absl::optional`

Reply:
done, looks much cleaner now, thanks !